include/onnxruntime/core/framework/data_types.h

```diff
@@ -694,6 +700,59 @@ class NonOnnxType : public DataTypeImpl {
   NonOnnxType() = default;
 };
 
+class PrimitiveDataTypeBase : public DataTypeImpl {
+ public:
+  bool IsCompatible(const ONNX_NAMESPACE::TypeProto&) const override {
+    return false;
+  }
+
+  const PrimitiveDataTypeBase* AsPrimitiveDataType() const override final {
+    return this;
+  }
+
+  const ONNX_NAMESPACE::TypeProto* GetTypeProto() const final {
+    return nullptr;
+  }
+
+  int GetTensorElementType() const {
```

Comment:
The name of this class doesn't seem to do anything with tensors; however it's methods indicate otherwise. For e.g. this method and also set_data_type which is valid for TensorProto only. Should these classes be called TensorElementType or TensorProtoDataType (to match the proto so that it's clearer)?

Reply:
The reason is the classes themselves can also be used for other purposes. They simply represent primitives' types. That specific method, however, returns a tensor specific enum integer describing how it is used specifically in tensors. We can attach other information for other purposes.

---
In reply to: [341770729](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r341770729) [](ancestors = 341770729)

Reply:
If the class simply represents primitive types, having a method called 'GetTensorElementType' seems very specific to a use case, and simply GetDataType may be a better method name . If you need something to provide the context of 'this is the element type in the tensor' that feels like it belongs outside of this class.

---
In reply to: [341783047](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r341783047) [](ancestors = 341783047,341770729)